src/syntax/syntax.rkt

```diff
@@ -111,46 +154,100 @@
 (define (no-complex fun)
   (λ xs
      (define res (apply fun xs))
-     (if (real? res)
-       res
-       +nan.0)))
+     (if (real? res) res +nan.0)))
 
 (define (default-nonffi . args)
   (raise
    (make-exn:fail:unsupported
     (format "couldn't find ~a and no default implementation defined" 'operator)
     (current-continuation-marks))))
 
-(define-operator (+ real real) real 
+(define parametric-operators (make-hash))
+(define parametric-operators-reverse (make-hash))
+
+(define (get-parametric-operator name actual-types)
+  (for/or ([sig (hash-ref parametric-operators name)])
+    (match-define (list* true-name rtype atypes) sig)
+    (and
+     (if (symbol? atypes)
+         (andmap (curry equal? atypes) actual-types)
+         (equal? atypes actual-types))
+     (cons true-name rtype))))
+
+(define (declare-parametric-operator! name op inputs output)
+  (hash-update! parametric-operators name (curry cons (list* op output inputs)) '())
+  (hash-set! parametric-operators-reverse op name))
+
+(define (operator-info* op field)
+  (cond
+   [(hash-has-key? parametric-operators op)
+    (let ([val (first (hash-ref parametric-operators op))])
+      (operator-info (car val) field))]
+   [else (operator-info op field)]))
```

Comment:
Generally suspicious of helpers like this. Where is it used, and can we move the logic there?

Also, are there changes to the parametric operator code, or was it just moved up here? Good idea to move it, just wondering.

Reply:
It's only used to get the expected number of arguments for an unparameterized operator. I suppose I could make a procedure just for that case since it's the only reason you would use `operator-info*`.

Reply:
Ah, ok. Honestly I'd just inline it where it is used. But we can also write a `parametric-operator-argnum` function too.

I think the ultimate goal is to have unparameterized operators exist somewhere other than keys in the `parametric-operators` hash, and we'll query that alternative data structure somewhere. So my goal here is to make it as easy as possible to make that change down the road. `operator*-info` doesn't make too clear which kind of operator it wants and why.